cro-clib/src/basic_wallet.rs

```diff
@@ -0,0 +1,166 @@
+    let address = StakedStateAddress::BasicRedeem(RedeemAddress::from(&public));
+
+    match address {
+        StakedStateAddress::BasicRedeem(redeem) => {
+            assert!(20 == redeem.0.len());
+            let raw = redeem.to_vec();
+
+            let ret = CroAddress {
+                privatekey: private,
+                publickey: public,
+                raw,
+                address: address.to_string(),
+            };
+            let address_box = Box::new(ret);
+            ptr::write(address_out, Box::into_raw(address_box));
+            CroResult::success()
+        }
+    }
+}
+
+/// create staking address
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_basic_create_staking_address(
+    address_out: *mut CroAddressPtr,
+) -> CroResult {
+    let private = PrivateKey::new().expect("get private key");
+    do_cro_basic_create_staking_address(address_out, private)
+}
+
+/// restore staking address
+/// input_length: maximum size of input
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_basic_restore_staking_address(
+    address_out: *mut CroAddressPtr,
+    input: *const u8,
+    input_length: u32,
+) -> CroResult {
+    assert!(32 == input_length);
+    let array: &[u8] = std::slice::from_raw_parts(input, input_length as usize);
+    match PrivateKey::deserialize_from(array) {
+        Ok(deserialized) => do_cro_basic_create_staking_address(address_out, deserialized),
+        Err(_) => CroResult::fail(),
+    }
+}
+
+/// # Safety
+unsafe extern "C" fn do_cro_basic_create_transfer_address(
+    address_out: *mut CroAddressPtr,
+    private: PrivateKey,
+) -> CroResult {
+    let public = PublicKey::from(&private);
+
+    let public_keys = vec![public.clone()];
+    let multi_sig_address = MultiSigAddress::new(public_keys, public.clone(), 1).unwrap();
+
+    let address: ExtendedAddr = multi_sig_address.into();
+
+    match address {
+        ExtendedAddr::OrTree(hash) => {
+            let raw = hash.to_vec();
+            // this is H256 hash
+            assert!(32 == raw.len());
+
+            let ret = CroAddress {
+                privatekey: private,
+                publickey: public,
+                raw,
+                address: address.to_string(),
+            };
+            let address_box = Box::new(ret);
+            ptr::write(address_out, Box::into_raw(address_box));
+
+            CroResult::success()
+        }
+    }
+}
+
+/// create staking address
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_basic_create_transfer_address(
+    address_out: *mut CroAddressPtr,
+) -> CroResult {
+    let private = PrivateKey::new().expect("get private key");
+    do_cro_basic_create_transfer_address(address_out, private)
+}
+
+/// restore staking address
+/// input_length: maximum size of input
+#[no_mangle]
+/// # Safety
+pub unsafe extern "C" fn cro_basic_restore_transfer_address(
+    address_out: *mut CroAddressPtr,
+    input: *const u8,
+    input_length: u32,
```

Comment:
does it need length argument if it only can be 32?

Reply:
i'll remove it , because it's always 32 bytes

Reply:
i'll remove it